### ISSUE TITLE
Priorities: new parameters

### ISSUE THREAD

**Issue by kixelated** [httpwg/http-extensions#1733]
One of the goals of this document is to define an extensible way of adding new priority parameters, including a registry. I've started on a draft that would define a new priority parameter `order`.

However, I'm running into some issues when `order` and `urgency` are used together. This seem inevitable because the server does not advertise it's capabilities, so my application is forced to send both if it wants _some_ form of prioritization.

There's a few options:

1. Define a backwards compatible way of supporting `urgency` alongside `order`. This is very difficult because they can cause wildly different behavior.
2. Ignore `urgency` when `order` used. This would mean that new priority parameters would effectively deprecate older parameters.
3. Advertise server capabilities out-of-band and only send one parameter.

**Comment by kixelated** [comment]
Yeah I don't really have any good recommendations. I like the simplicity of a `Priority` header, but I think it's going to become a mess as new strategies are added.

My application does need to know if prioritization is supported before it can issues requests in parallel. I'm thinking some form of hop-by-hop capabilities exchange (option 3) accomplishes that while also addressing this issue, but it's just so much extra complexity.

I think the right action is to make these initial priority parameters powerful enough to cover most use-cases, removing the need to establish new parameters. I know that goes against the spirit of the draft.

**Comment by kazuho** [comment]
Yeah, ordering is about prioritizing requests relative to others, while Extensible Priorities focus on absolute values. This has been an deliberate choice; please refer to [section 2](https://httpwg.org/http-extensions/draft-ietf-httpbis-priority.html#name-motivation-for-replacing-rf).

While it might be possible to use Priority Parameters to signal ordering, as @kixelated points out, there has to be negotiation or the signal should only be sent using frames. We can discuss that as an extension, but I would say that that is out of scope of this draft.

**Comment by kixelated** [comment]
My apologies for not being on the list sooner. I've had the email open to subscribe multiple times now but never actually hit send.

I've thought about prioritization a lot, although probably not as much as you guys. The [CloudFlare blog](https://blog.cloudflare.com/better-http-2-prioritization-for-a-faster-web/) has such a great write-up even if HTTP/2 prioritization did not catch on.

The success of a prioritization scheme depends on the user experience during congestion. For a web browser, there's a bunch of little milestones, such as being able to display text, apply stylesheets, run scripts, load images, etc. The sooner you can get to these milestones, the better the user experience. This means spending all available bandwidth on the first milestone, spending the remaining bandwidth on the second milestone, spending the remaining bandwidth on the third milestone, etc.

I haven't worked on a browser, but I don't quite see how `urgency` accomplishes this without stricter guarantees. If early milestone resources are not being prioritized, and on each hop, then requesting later milestone resources in parallel may hurt the user experience.

For example, simultaneously requesting images with `u=3` and scripts with `u=5`. If there's congestion on any hop and that sender does not send `u=5` before `u=3` (for whatever reason), then it may be a better user experience to request them sequentially instead.

I don't think that HTTP live streaming is a fundamentally different problem. You can think of each HLS segment as a "milestone", and indeed it's currently better to request these segments sequentially than have them fight for bandwidth in parallel. In fact I think HTTP live streaming is the ideal prioritization use-case because resources are massive, data transfer is constant, congestion is inevitable, and the user experience is terrible if a resource does not arrive in time (buffering).

**Comment by LPardue** [comment]
No need to apologise, we all suffer from too many list syndrome :-) 

> The success of a prioritization scheme depends on the user experience during congestion. 

If the use of the term congestion here is implying transport only aspects I disagree. It's entirely possible to have a bad user experience with a perfect signalling scheme, that a server is known to implement exactly as defined because, for example, the seek time to a file on a hard drive is slow, or there is a long RTT to a the server, or the DNS resolution for the origin that the server is on is failing, or etc etc. All of those things can happen when there is no transport layer congestion. Perhaps you mean availability of data to serve?

> For a web browser, there's a bunch of little milestones, such as being able to display text, apply stylesheets, run scripts, load images, etc. The sooner you can get to these milestones, the better the user experience. This means spending all available bandwidth on the first milestone, spending the remaining bandwidth on the second milestone, spending the remaining bandwidth on the third milestone, etc.

That depends on the precise makeup of a webpage. Some resource need to be served in full before they are useful. Some resources can be processed incrementally. There's a lot of nuances here. The challenge, that you correctly point out, is to find the most appropriate use of available bandwidth to multiplex response and to deliver the critical parts of resources with minimal delay from the ideal edge of their availability. Where resources are not available to be served against the ideal edge, then a server has to make a judgement on whether to commit bandwidth to some resources. That bandwidth cannot be easily reclaimed if there is a high BDP or large buffers between the server and client.

We should also not make the assumption that all sections of a resource are equally important. For instance, with progressive images the early bytes are more important than the later bytes . This is explored more in https://blog.cloudflare.com/parallel-streaming-of-progressive-images/, where we use a response header to inform the intermediary of precise byte boundaries only the image server knows. One might argue a client could reprioritize after the byte boundaries are crossed, but the reality is that responsiveness across RTTs is probably too slow when the bandwidths these days are so large. Fine margins here. The `cf-priority-change` header was designed before `Priority` but the design could be ported over as a new priority parameter. This demonstrates much more flexibility and control than is possible with RFC 7540 stream priorities. 

> I haven't worked on a browser, but I don't quite see how urgency accomplishes this without stricter guarantees. If early milestone resources are not being prioritized, and on each hop, then requesting later milestone resources in parallel may hurt the user experience.

> For example, simultaneously requesting images with u=3 and scripts with u=5. If there's congestion on any hop and that sender does not send u=5 before u=3 (for whatever reason), then it may be a better user experience to request them sequentially instead.

A server is always making prioritization decisions whether there is a priority signal (i.e the priority header) or not. 

In your example, the client is preferring non-incremental responses for images over non-incremental respones for scripts. It is recommended that non-incremental responses are served in the order they were requested. So if a client issues on HTTP/3

stream 0: /image1.png
stream 4: /script1.js
stream 8: /image2.png
stream 12: /script2.js

If all those resources are available to a server that follows the spec guidance, the order of response data would be: image1.png, image2.png, script1.js, script2.js.

If image2.png was not readily available, the order of response data would be: image1.png, script1.js, script2.js ... sometime later image2.png.

If image2.png is slightly delayed, a server could prempt things. Such as image1.png, half of script1.js, image2.png, half of script1.js, script2.js

If some or all of those requests were made with an `incremental` flag. The server would make different choices.

> If I already know that I'm requesting a resource from CloudFlare or Fastly and they will prioritize in a particular manner, then there's no problem. My concern is that a generic video player or browser doesn't know this information ahead of time.

I don't follow. The web doesn't operate in a vacuum. Players or browser operate on behalf of a user, who is accessing an application or service. That is likely to have been designed by somebody with an interest in the application and its behavior. That somebody should know where their resources are being served from. Speaking from experience of a video player service that operated a multi-CDN solution, the use of DNS names is a simple and powerful method to know where responses and coming from and the types of implementations that are powering them. 

The worst thing that can happen if a server doesn't prioritie well is that application measures of success will score badly. Users may notice, or they may not care. Operators may notice, or they may not care. People who do care may notice and make a page that names and shames bad performing implementations - https://github.com/andydavies/http2-prioritization-issues. You can see on that test case a depressingly large number of cloud hosting services did not do very well and, to my knowledge, few ever remediated the issues despite customers filing compliants or feature requests. I hate to pick on anyone but Amazon CloudFront is a major example of this problem - what would it take to make them change? I don't think a nigh-impossible to achieve requirement in a specification is going to shift the needle.

> I don't think that HTTP live streaming is a fundamentally different problem. You can think of each HLS segment as a "milestone", and indeed it's currently better to request these segments sequentially than have them fight for bandwidth in parallel. In fact I think HTTP live streaming is the ideal prioritization use-case because resources are massive, data transfer is constant, congestion is inevitable, and the user experience is terrible if a resource does not arrive in time (buffering).

I think this depends on the latency requirements of the stream. Speaking from experience, traditional HLS or DASH players build up a multi-segment (multi-second) buffer in order to avoid buffer underflow and playback disruption. What that means is that clients will get into a pattern of periodic traffic bursts when fetchin the next segment to buffer and then have periods cessation. The bandwidth available to clients is typically massive in comparison to the size of segment being downloaded.

Stream start and seek times are important and can be different to above. There you might need to focus on quick delivery of the thing that needs to be played over something else. That's where an inflight request for something you no longer want can steal bandwidth from the thing you do. And since this is an unpredictable user action, the only recourse is stream cancellation or reprioritization. And then you get back into the realm of delayed response, RTTs and BDP.

Traditionally, with segment sizes have been so large (in the time domain) that it's easy for a client to manage the ordering of things. Yes you could in HTTP/2, for instance with a VOD stream, just request all of the segments at once and punt the problem of ordering to the server.  But then that's not going to be very ABR-like. So client's should only have a few requests in flight to not over commit their ABR algorithm. In something like DASH where audio and video segments are not multiplexed at the encoder, the client is tasked with muxing them at synchronization points. Which audio or video segment comes first, or how they compete for bandwidth is meaningless because they both need to arrive in order to make a meaningful presentation of them. If a client is only issuing a few requests in-flight at a time, the problem of order is trivial. You don't need any additional signal, just look at the order.

Low latency ABR video reduces the length of segments (or may even expose sub-segment primitives). This massively reduces the size of buffers and the margin for error because RTT starts to approach segment length. The needs are a bit different hear. 2 years ago, large parts of the industry were told that the only way to satisfy the needs of low latency video stream was to run HTTP/2 server push through the entirety of a video distribution chain, see https://mux.com/blog/the-community-gave-us-low-latency-live-streaming-then-apple-took-it-away/. And folks soon realized that placing such requirements on implementations isn't very practical, leading to tit being replaced by something that was https://mux.com/blog/low-latency-hls-part-2/. "Blocking requests" can be made before a segment is available and servers won't reject them instead they'll hold them and server them when the resource is available) this help eliminate RTTs that might cause interruptions. Blocking requests can be made in batches and here the ordering of serving is important, requests from the start of the batch through to the end of the batch. I'd argue that this is a no brainer for a server to handle without any additional priority signal. But if we want one, it is easily supported by sending all requests with an extensible priority of `u=3` - a server that follows the guidance in the spec will serve these sequentially in the order they were issued.

From the other issue it sounded like you wanted something different, where servers employ a reverse order. You could achieve that by an implicit-oriented boolean parameter like `reverse-order=?1`, rather than trying to explicitly order each and every request. I could be misunderstanding though.

**Comment by kixelated** [comment]
> A server is always making prioritization decisions whether there is a priority signal (i.e the priority header) or not.

Ah okay, I think this is the source of some of our disagreement. I've been under the assumption that this client header is the only prioritization parameter. But I understand from your vantage point, CDN customers also use their own business logic to prioritize on the server.

I would like to define a video protocol, and as you pointed out with LL-HLS, it needs to be simple enough to adopt. The protocol depends on some way of defining a priority per segment, either a query parameter or a header or perhaps some server-side logic.

In my case, I don't think it's feasible to ask providers to use and configure a CDN such that it prioritizes the video content according to specific rules. Imagine if LL-HLS was only supported by a handful of CDNs; the community would lash out even harder than the HTTP/2 requirement. 

The prioritization logic I need is very simple; deliver higher priority response first when possible. This draft is sooooo close but it just barely misses the mark.
